_posts/2021-04-22-torch-for-optimization/torch-for-optimization.Rmd

```diff
@@ -0,0 +1,640 @@
+
+```{r}
+num_iterations <- 1000
+
+# fraction of the gradient to subtract 
+lr <- 0.01
+
+# function input (x1,x2)
+# this is the tensor w.r.t. which we'll have torch compute the gradient
+x_star <- torch_tensor(c(-1, 1), requires_grad = TRUE)
+
+for (i in 1:num_iterations) {
+
+  if (i %% 100 == 0) cat("Iteration: ", i, "\n")
+
+  # call function
+  value <- rosenbrock(x_star)
+  if (i %% 100 == 0) cat("Value is: ", as.numeric(value), "\n")
+
+  # compute gradient of value w.r.t. params
+  value$backward()
+  if (i %% 100 == 0) cat("Gradient is: ", as.matrix(x_star$grad), "\n\n")
+
+  # manual update
+  with_no_grad({
+    x_star$sub_(lr * x_star$grad)
+    x_star$grad$zero_()
+  })
+}
+```
+
+    Iteration:  100 
+    Value is:  0.3502924 
+    Gradient is:  -0.667685 -0.5771312 
+
+    Iteration:  200 
+    Value is:  0.07398106 
+    Gradient is:  -0.1603189 -0.2532476 
+
+    ...
+    ...
+
+    Iteration:  900 
+    Value is:  0.0001532408 
+    Gradient is:  -0.004811743 -0.009894371 
+
+    Iteration:  1000 
+    Value is:  6.962555e-05 
+    Gradient is:  -0.003222887 -0.006653666 
+
+While this works, it really serves to illustrate the principle only. With `torch` providing a bunch of proven optimization algorithms, there is no need for us to manually compute the candidate $\mathbf{x}$ values.
+
+## Function minimization with `torch` optimizers
+
+Instead, we let a `torch` optimizer update the candidate $\mathbf{x}$ for us. Habitually, our first try is *Adam*.
+
+### Adam (?)
+
+With Adam, optimization proceeds a lot faster. Truth be told, though, choosing a good learning rate *still* takes non-negligeable experimentation. (Try the default learning rate, 0.001, for comparison.)
+
+```{r}
+num_iterations <- 100
+
+x_star <- torch_tensor(c(-1, 1), requires_grad = TRUE)
+
+lr <- 1
+optimizer <- optim_adam(x_star, lr)
+
+for (i in 1:num_iterations) {
+  
+  if (i %% 10 == 0) cat("Iteration: ", i, "\n")
+  
+  optimizer$zero_grad()
+  value <- rosenbrock(x_star)
+  if (i %% 10 == 0) cat("Value is: ", as.numeric(value), "\n")
+  
+  value$backward()
+  optimizer$step()
+  
+  if (i %% 10 == 0) cat("Gradient is: ", as.matrix(x_star$grad), "\n\n")
+  
+}
+
+```
+
+    Iteration:  10 
+    Value is:  0.8559565 
+    Gradient is:  -1.732036 -0.5898831 
+
+    Iteration:  20 
+    Value is:  0.1282992 
+    Gradient is:  -3.22681 1.577383 
+
+    ...
+    ...
+
+    Iteration:  90 
+    Value is:  4.003079e-05 
+    Gradient is:  -0.05383469 0.02346456 
+
+    Iteration:  100 
+    Value is:  6.937736e-05 
+    Gradient is:  -0.003240437 -0.006630421 
+
+It took us about a hundred iterations to arrive at a decent value. This is a lot faster than the manual approach above, but still quite a lot. Luckily, further improvements are possible.
+
+### L-BFGS (!)
```

Comment:
I now see that it's quite a mouthful to spell this one out. You could have it as a hoverable footnote earlier, perhaps?